Added data collator for permutation (XLNet) language modeling and related calls

shngt opened the pull request:
Added `DataCollatorForPermutationLanguageModeling` in `data/data_collator.py` to return necessary inputs (applies masking and generates revelant tensors input_ids, perm_mask, target_mask and labels as per https://github.com/zihangdai/xlnet/blob/master/data_utils.py) for language modeling training with XLNetLMHeadModel. Also added related arguments, logic and calls in `examples/language-modeling/run_language_modeling.py`. Defined a separate `--plm_probability` flag for its use.

Also looked into CTRL - it uses a CLM loss just like GPT and GPT-2, so should work out of the box with this script (provided `past` is taken care of similar to `mems` for XLNet). Added a few words in the comments to reflect this.

Changed calls and imports appropriately.

Resolves: #4739, #2008 (partially)

shngt committed:
Added data collator for XLNet language modeling and related calls

Added DataCollatorForXLNetLanguageModeling in data/data_collator.py
to generate necessary inputs for language modeling training with
XLNetLMHeadModel. Also added related arguments, logic and calls in
examples/language-modeling/run_language_modeling.py.

Resolves: #4739, #2008 (partially)

shngt committed:
Changed name to `DataCollatorForPermutationLanguageModeling`

Changed the name of `DataCollatorForXLNetLanguageModeling` to the more general `DataCollatorForPermutationLanguageModelling`.
Removed the `--mlm` flag requirement for the new collator and defined a separate `--plm_probability` flag for its use.
CTRL uses a CLM loss just like GPT and GPT-2, so should work out of the box with this script (provided `past` is taken care of
similar to `mems` for XLNet).
Changed calls and imports appropriately.

shngt commented:
@patrickvonplaten Please review

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
```
It should be `plm_probability` here no?

patrickvonplaten replied:
Copy past comment 2) from algo here

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+
+        if self.tokenizer.mask_token is None:
+            raise ValueError(
+                "This tokenizer does not have a mask token which is necessary for masked language modeling. Remove the --mlm flag if you want to use this tokenizer."
```
Should the comment be updated for plm instead of mlm?

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
+                start = cur_len + torch.randint(ctx_size - n + 1, (1,)).item()
+                end = start + 1
```
Can't all this be simplified to something like: `end_idx = start + max(1, min(max_len - start, n))` ?

patrickvonplaten replied:
Actually I think it's just `end_idx = min(max_len, start + span_len_tokens_to_predict)` no?

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+                while end < max_len and end < start + n:
+                    end += 1
+                masked_indices[i, start:end] = 1
+                cur_len += ctx_size
```
Is that equivalent to this line in the original code? https://github.com/zihangdai/xlnet/blob/bbaa3a6fa0b3a2ee694e8cf66167434f9eca9660/data_utils.py#L381 ?

patrickvonplaten replied:
But it makes sense!

shngt replied:
Yes, I think so.

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            # the i-th label corresponds to the i-th token.
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
```
Can we give `n` a better name => n_gram_length . 
Also in the official code `n_gram_length` is not sampled uniformely: https://github.com/zihangdai/xlnet/blob/bbaa3a6fa0b3a2ee694e8cf66167434f9eca9660/data_utils.py#L351

Do you think this is not very important?

patrickvonplaten replied:
Ok, I re-read the paper and looked into the code again -> I think a good name for n would actually be `span_len_tokens_to_predict`. And it would be nice if we can call `self.max_gram` also self.max_span_len_tokens_to_predict`. As far as I understand now, the actually number of tokens that will be predicted is independent of `plm_probability` and only depends on `self.max_span_len_tokens_to_predict` (or `self.max_gram`). So I think `plm_probability` is a bit confusing here. As far as I understand `plm_probability` only defines the **start** index when tokens shall be predicted. I think this is also not very clearly written in the paper, but just looking at the official code, to me it seems that `plm_probability` (or `mask_beta / mask_alpha` from the official code) does not influence the actualy number of tokens that shall be predicted as does `mlm_probability`. So I think we should not call it plm_probability but more like `pred_context_len_beta` or `pred_context_len_prob`.

patrickvonplaten replied:
Copy past comment 1) from algorithm above before the `span_len` is sampled.

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+        # Mask indicating non-functional tokens, where functional tokens are [SEP], [CLS], padding, etc.
+        non_func_mask = ~(padding_mask & special_tokens_mask)
+
+        inputs[masked_indices] = self.tokenizer.convert_tokens_to_ids(self.tokenizer.mask_token)
```
better to use `self.tokenizer.mask_token_id` instead of calling `convert_tokens_to_ids`

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+        inputs[masked_indices] = self.tokenizer.convert_tokens_to_ids(self.tokenizer.mask_token)
+        labels[~masked_indices] = -100  # We only compute loss on masked tokens
+
+        perm_mask = torch.zeros((labels.size(0), labels.size(1), labels.size(1)), dtype=torch.float)
```
`torch.float32` here for clarity

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            # Generate permutation indices
+            # Maximum permutation length has to be less than or equal reused sequence length,
+            # otherwise leakage will occur. We assume that reused length is half of sequence length
+            # and equality is enforced. This requires that sequence length be even.
```
"requires that sequence length is even" => can we check that with an `assert` statement?

patrickvonplaten replied:
Would be nice to have an assert statement here that throws an error if the sequence length is not even

patrickvonplaten commented on examples/language-modeling/run_language_modeling.py:
```diff
@@ -101,6 +102,13 @@ class DataTrainingArguments:
         default=0.15, metadata={"help": "Ratio of tokens to mask for masked language modeling loss"}
     )
+    plm_probability: float = field(
+        default=0.15,
```
I think default is set to 1/6 in the official code no? => maybe do the same here?

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+    """
+
+    tokenizer: PreTrainedTokenizer
+    plm_probability: float = 0.15
```
I think default is set to 1/6 in the official code no? => maybe do the same here?

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The mask is determined by iteratively performing the following steps until the end of the sequence:
+            1. Sample an n-gram length (n) from [1, ..., max_gram] (L in the paper)
```
`span_length` maybe instead of `n-gram`

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
```
Also usually we write out small words like `ctx` => `context_length`

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The mask is determined by iteratively performing the following steps until the end of the sequence:
```
This is super useful!

Some improvement suggestions:
`The masked tokens to be predicted are determined by .... untitl the end of the sequence (cur_len >= max_length)

Start at `cur_len = 0`

1. ...
2. ... of length `context_length = span_length / plm_probability`
3. ... Sample a starting point `start_index` from the interval `[cur_len, cur_len + context_length - span_length]` and mask tokens `start_index: start_index: span_length`
4. ... Set cur_len = cur_len + context_length

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
+                start = cur_len + torch.randint(ctx_size - n + 1, (1,)).item()
```
copy past comment 3) from algo here

patrickvonplaten replied:
also can we use `start_idx` instead?

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+            # the i-th label corresponds to the i-th token.
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
```
Can we call `n` `span_length` and `self.max_gram` => `self.max_span_length` or even self.max_span_length_tokens_to_predict` ? I think this would make the code much clearer here

patrickvonplaten commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+                end = start + 1
+                while end < max_len and end < start + n:
+                    end += 1
+                masked_indices[i, start:end] = 1
```
Seems a bit weird how `end` is calculated => I think just doing:
```
masked_indices[i, start_idx: start_idx + span_length] = 1
```
should be enough no? If `start_idx + span_length` is bigger than `max_length`, no error will be thrown in Python as far as I know.

shngt replied:
Yes, it seems so (I was surprised to see this!). Would you prefer `start_idx + span_length` or `min(start_idx + span_length, max_len)`?

patrickvonplaten replied:
just `start_idx + span_length` should be fine I think :-)

patrickvonplaten commented:
Hey @shngt - Thanks a mille for your PR. Pretraining for XLNet is not easy at all, so this is a super valuable PR :-) 

Just re-read the paper and looked into the official xlnet code -> I think I finally understand now how it works exactly :D 
Because it's not at all straight-forward to understand what happens there, it would be great if we can try to make the code as understandable as possible -> that's why I left so many comments mainly on variable names.

Also, we should add a test to `test_trainer.py` to next to the other data collators to test this one :-) 

Looking forward to merge this soon :-)

shngt committed:
Added detailed comments, changed variable names

Added more detailed comments to `DataCollatorForPermutationLanguageModeling` in `data/data_collator.py` to explain working. Also cleaned up variable names and made them more informative.

shngt committed:
Added tests for new data collator

Added tests in `tests/test_trainer.py` for DataCollatorForPermutationLanguageModeling based on those in DataCollatorForLanguageModeling. A specific test has been added to check for odd-length sequences.

shngt commented:
@patrickvonplaten I've added more detailed comments and fixed the naming issues. I also added a few tests based on what I could see in `tests/test_trainer.py`. Let me know what you think :)

patrickvonplaten commented:
Awesome job @shngt ! This looks very clean now :-) Good to merge IMO.
Pinging @julien-c @LysandreJik.

sgugger commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,127 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The masked tokens to be predicted for a particular sequence are determined by the following algorithm:
+            0. Start from the beginning of the sequence by setting `cur_len = 0` (number of tokens processed so far).
```
Here and in all docstrings, please use sphinx restructured text syntax (which is different from Markdown): putting an object between backticks just render it in italics. I think you want double-backticks or a syntax like ``obj:`object` ``.
So here either ` ``cur_len = 0`` ` or ``obj:`cur_len=0` `` and same in the rest of this docstring.

shngt replied:
@sgugger Yes, I think changing it to double backticks should work. I noticed that the docstring in `default_data_collator` also uses single backticks - would you like me to change that as well? Also, is this formatting required in ordinary comments?

sgugger replied:
This is for proper rendering on the online doc, so it does not apply to ordinary comments. Yes please for `default_data_collator`, I haven't gotten to fix this module for the docs yet (not even sure it's documented online yet to be honest).

shngt replied:
Okay. I've left the single backticks in for the comments to help distinguish variable names and so on from regular text. Let me know if you'd like me to remove those as well.

sgugger replied:
LGTM, thanks!

sgugger commented on src/transformers/data/data_collator.py:
```diff
@@ -134,3 +134,127 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+        perm_mask = torch.zeros((labels.size(0), labels.size(1), labels.size(1)), dtype=torch.float32)
+
+        for i in range(labels.size(0)):
+            """
```
Should be a comment?

shngt replied:
No problem, will fix

sgugger reviewed (approved):
LGTM, just a few things to fix for the doc and one nit.

shngt committed:
Fixed styling issues

LysandreJik reviewed (approved):
Very clean, and great comments! Thanks a lot!

patrickvonplaten commented:
Pinging @thomwolf and @julien-c for notification and merging - Great job @shngt!

krannnn commented:
Impressive work, I was following this from the beginning.

Can we continue training XLNet now on domain-specific corpus ? and when will this merge be available to use? 

Thanks

patrickvonplaten commented:
PR should be ready for use as of now, if you install from master :-)

julien-c commented:
Great work @shngt, this is very nice.

WenjieDu commented:
I've raised an issue here #22435 because I have a question about this `DataCollatorForPermutationLanguageModeling`. Looking forward to your guys' response. Thank you!
